util/persistence.py

```diff
@@ -0,0 +1,242 @@
+        '''
+        Deletes files of a file-set.
+        '''
+        for filename in filenames:
+            log.step('Removing file "%s"...' % filename)
+            os.remove(filename)
+
+    def _prune_and_save(self, session, epoch_index, global_index):
+        '''
+        Removes outdated checkpoint files and stores a new checkpoint.
+        '''
+        keep = self.keep_n_inters if epoch_index == 0 else self.keep_n_epochs
+        assert keep > 0
+        checkpoints = []
+        for checkpoint in self._get_checkpoints():
+            # if checkpoint's global index is greater than this one,
+            if checkpoint[1] > global_index:
+                # it is on an outdated branch of training history and will be deleted
+                log.debug('Removing files of checkpoint "%s", as the global index %d is greater than the new one (%d)...' % \
+                    (self._get_filename(filenames), checkpoint[1], global_index))
+                self._delete_files(filenames)
+            else:
+                # for the next steps we only keep checkpoints of the current
+                # checkpoint family (either epoch ones or intermediate ones)
+                if (epoch_index == 0 and checkpoint[0] == 0) or (epoch_index > 0 and checkpoint[0] > 0):
+                    checkpoints.append(checkpoint)
+        # removing checkpoint files within current checkpoint family that fall off the keep range
+        for _, _, filenames in checkpoints[:-(keep - 1)]:
+            log.debug('Removing files of "%s", as only %d checkpoints should be kept...' % \
+                (self._get_filename(filenames), keep))
+            self._delete_files(filenames)
+        # generating checkpoint filename
+        filename = '%s_%d.ckpt' % (('epoch-%d' % epoch_index) if epoch_index > 0 else 'inter', global_index)
+        filename = os.path.join(self.checkpoint_dir, filename)
+        log.info('Writing checkpoint "%s"...' % filename)
+        # finally: saving the new checkpoint
+        self.saver.save(session, filename, write_state=False)
+
+    def _init(self, session):
+        '''
+        Initializes the session.
+        '''
+        log.info('No checkpoint to load - initializing new model...')
+        session.run(self.init_op)
+
+    def _restore(self, session, filename):
+        '''
+        Restores a checkpoint file-set.
+        '''
+        log.info('Restoring checkpoint "%s"...' % filename)
+        self.saver.restore(session, filename)
+
+    def start(self, session):
+        '''
+        Starts a training session by either restoring the right checkpoint
+        or by initializing the model.
+        Should be called at the beginning of a training session.
+        '''
+        if not self.checkpoint_dir:
+            self._init(session)
+            return
+        checkpoint = None
+        # collect available checkpoints
+        checkpoints = self._get_checkpoints()
+        if self.load == 'recent':
```

Comment:
The more I read all this code, the more I think it would at least help to have some checks, maybe in `__init__()` against values of `self.load`. Definition of all the accepted values are buried in the code

Reply:
Unfortunately not possible, as the final fall-back is to load a file at path `self.load`. And checking for the existence of a file during object construction is bad style. The values are documented as part of `__init__`.